README
version 0.9923_03
==================================

Object oriented versions for all Perl releases from 5.6.2 onward.  Replaces
the core version code for all Perl releases from 5.10.0 onwards.

Although this code has been fully integrated into the latest Perl release
itself (e.g. bleadperl), the CPAN release is considered upstream.  The source
code itself is available from two mirrors:

    https://bitbucket.org/jpeacock/version (primary)
    https://github.com/JohnPeacock/version.git (secondary)

and a bug queue here:

   https://rt.cpan.org/Dist/Display.html?Queue=version

Patches welcomed.

cperl support is maintained here:

      https://github.com/rurban/version

IMPORTANT NOTE - 0.9906 is the last release that fully supports Perl
releases prior to 5.6.2.  It is just too much trouble to continue to
fight to maintain full backward compatibility.

YOU SHOULD INSTALL THIS RELEASE WHEN USING VERSION OBJECTS IN PERL 5.10.0.
There was a bug caused by objects serialized by YAML that will cause Perl
5.10.0 to segfault.  As long as some module calls 'use version', then the
replacement code will take effect and Perl will not crash.

Please read the POD documentation for usage/details.  See the CHANGES file
for full details of all changes to the module behavior.

INSTALLATION

To install this module type the following:

   $ perl Makefile.PL
   $ make
   $ make test
   # make install

This release includes an optional pure Perl implementation (in case
you don't have a C-compiler or if some reason you want your code to be
much slower).  You can test it by replacing the first line above with:

   $ perl Makefile.PL --perl_only

and it will install the Perl only version.

DEPENDENCIES

The same C compiler used to build Perl (or not).


COPYRIGHT AND LICENCE

This module can be distributed under the same terms as Perl.

Copyright (C) 2004-2017 John Peacock



CHANGES
Major changes for Perl module version

This is not a complete list of changes.  See repository for full details:

    https://github.com/JohnPeacock/version.pm.git

0.9924_02 - 2018-10-13  (rurban)

  * merged with cperl
  * more 5.6 fixes

0.9924 - 2018-04-19
  * Revise CPAN metadata to advertise correct repo
	Patch from Ed <ej_zg@hotmail.com>

0.9923_02 - 2017-09-07  (rurban)

  * merged with cperl, use vars => our.

0.9923 - 2018-04-15
  * Add back spaces for picky C++ compilers 
	Patch from Karl Williamson <khw@cpan.org>

0.9922 - 2018-04-12
  * Switch master repo to github

0.9921 - 2018-04-11
  * Silence overly compulsive GCC 7 warning
	https://rt.cpan.org/Ticket/Display.html?id=123477

  * Improve docs regarding trailing zeros
	https://rt.cpan.org/Ticket/Display.html?id=122858

0.9920 - 2018-04-08

  * Backwards compatible locale handling
	https://rt.cpan.org/Ticket/Display.html?id=125042

0.9991_01 - 2017-04-08  (rurban)

  * merged cperl support

0.9919 - 2018-04-08

  * Core improvements to locale handling
	https://rt.cpan.org/Ticket/Display.html?id=124563

0.9918_01c - 2017-04-16  (rurban)

  * merged cperl support

0.9918 - 2017-04-15

  * Fix compiling under C++11
	https://rt.cpan.org/Ticket/Display.html?id=118846

  * Make version::regex variables public for better reuse
	https://rt.cpan.org/Ticket/Display.html?id=119669

  * Use rel2abs for require in tests to support 5.26.0 @INC
	https://rt.cpan.org/Ticket/Display.html?id=121148

  * Move change list to standard CHANGES out of README
	https://rt.cpan.org/Ticket/Display.html?id=119123

  * Better taint handling for Perl < 5.17.2 with tests
	https://rt.cpan.org/Ticket/Display.html?id=118087

0.9917_02c - 2016-10-29 (rurban)

  * add more 04strict_lax.t cperl tests
  * Update to 0.9917 lyon _ concensus
  * plus several cperl fixes

0.9917_01c - 2016-10-16 (rurban)

  * merged cperl support
  * parse v1.0c natively in toke.c
    See GH #211
    upgrade version::regex to parse the c suffix. (lax/strict)

  * not merged the version 0.9917 alpha/width removal,
  * just some '_' fixes.

0.9917 - 2016-05-29

  * Fix version::regex captures, resolves:
	https://rt.cpan.org/Ticket/Display.html?id=114712

0.9916 - 2016-03-18

  * Incorporate changes from bleadperl (destined to be 5.24.0)

0.9915 - 2016-03-05

  * Clarify details of Lyon Compromise:
      https://gist.github.com/dagolden/9559280

    - version comparision should be done irrespective of the presence of
      underscores in the string used to initialize the version object

    - underscore should no longer be used as a tuple separator in vstrings or
      vstring-like strings; vstrings are converted to tuples by splitting into
      *characters* (not bytes) and converting to codepoints; any elements after
      the first must be in the range 0-999

    - numify/normal should produce a standarized string representation without
      underscores

    - stringify should produce the best possible representation of the value
      used to initialize the version object; it should include underscores
      only if the initializing value was a non-vstring string.

    - floating point numbers used as initializers are converted to a decimal
      string form at the precision limit of the architecture; people will be
      warned about this in the documentation

0.9914 - 2016-02-18

  * Release non-alpha version to CPAN

0.9913_01 - 2016-02-17

  * Apply one additional changeset from leont to handle magic v-strings

0.9913 - 2016-02-16

  * Release non-alpha version to CPAN

0.9912_03 - 2016-01-21

  * Apply two pull requests from bitbucket:
    https://bitbucket.org/jpeacock/version/pull-requests/1/fix-test-failures-due-to-hard-coded/diff
    https://bitbucket.org/jpeacock/version/pull-requests/2/static-all-symbols/diff

0.9912_02 - 2016-01-03

  * Fix stupid mistake in locale tests, resolves:
	https://rt.cpan.org/Ticket/Display.html?id=110852

0.9912_01 - 2015-12-31

  * Apply patch to implement Lyon Consensus:
	https://gist.github.com/dagolden/9559280

  * Also resolve the following tickets:
	https://rt.cpan.org/Ticket/Display.html?id=101647
	https://rt.cpan.org/Ticket/Display.html?id=101841
	https://rt.cpan.org/Ticket/Display.html?id=105315
	https://rt.cpan.org/Ticket/Display.html?id=106782
	https://rt.cpan.org/Ticket/Display.html?id=107113
	https://rt.cpan.org/Ticket/Display.html?id=107114
	https://rt.cpan.org/Ticket/Display.html?id=98744
	https://rt.cpan.org/Ticket/Display.html?id=102272

0.9912 - 2015-01-20

  * Complete reverting the ill-conceived alpha->normal() code, resolves:

	https://rt.cpan.org/Ticket/Display.html?id=101632

  * Skip leading zeros when parsing dotted-decimal versions after the
    first element, resolves:

	https://rt.cpan.org/Ticket/Display.html?id=101628

  * Minor change to report the correct rt.cpan.org URL, resolves:

	https://rt.cpan.org/Public/Bug/Display.html?id=101600

0.9911 - 2015-01-17

  * Backout change to normal() and numify() because it breaks Module::Build

0.9910 - 2015-01-17

  * Ensure that calling either normal() or numify() on an alpha version
    is explicitly a lossy operation, and add a warning to that effect.  Forbid
    trailing bare decimal for dotted-decimal versions.  Change heuristic in vpp
    to better resolve v-string versions.  Various cleanups.

    Resolves:

	https://rt.cpan.org/Ticket/Display.html?id=93603
	https://rt.cpan.org/Ticket/Display.html?id=93715
	https://rt.cpan.org/Ticket/Display.html?id=93721
	https://rt.cpan.org/Ticket/Display.html?id=95896
	https://rt.cpan.org/Ticket/Display.html?id=96100
	https://rt.cpan.org/Ticket/Display.html?id=96620
	https://rt.cpan.org/Ticket/Display.html?id=96699

  * Patch from Dave Mitchell to rewrite the parser and avoid warnings
    from clang.  Resolves

	https://rt.cpan.org/Ticket/Display.html?id=101501

0.9909 - 2014-08-15

  * Compatibility release for Perl 5.20.1.  Apply two patches from bleadperl

vutil/vutil.h
/* This file is part of the "version" CPAN distribution.  Please avoid
   editing it in the perl core. */

#ifndef PERL_CORE
#  include "ppport.h"
#endif

/* The MUTABLE_*() macros cast pointers to the types shown, in such a way
 * (compiler permitting) that casting away const-ness will give a warning;
 * e.g.:
 *
 * const SV *sv = ...;
 * AV *av1 = (AV*)sv;        <== BAD:  the const has been silently cast away
 * AV *av2 = MUTABLE_AV(sv); <== GOOD: it may warn
 */

#ifndef MUTABLE_PTR
#  if defined(__GNUC__) && !defined(PERL_GCC_BRACE_GROUPS_FORBIDDEN)
#    define MUTABLE_PTR(p) ({ void *_p = (p); _p; })
#  else
#    define MUTABLE_PTR(p) ((void *) (p))
#  endif

#  define MUTABLE_AV(p)	((AV *)MUTABLE_PTR(p))
#  define MUTABLE_CV(p)	((CV *)MUTABLE_PTR(p))
#  define MUTABLE_GV(p)	((GV *)MUTABLE_PTR(p))
#  define MUTABLE_HV(p)	((HV *)MUTABLE_PTR(p))
#  define MUTABLE_IO(p)	((IO *)MUTABLE_PTR(p))
#  define MUTABLE_SV(p)	((SV *)MUTABLE_PTR(p))
#endif

#ifndef hv_existss
#  define hv_existss(hv, key) \
          hv_exists((hv), ("" key ""), (sizeof(key)-1))
#endif

#ifndef SvPVx_nolen_const
#  if defined(__GNUC__) && !defined(PERL_GCC_BRACE_GROUPS_FORBIDDEN)
#    define SvPVx_nolen_const(sv) ({SV *_sv = (sv); SvPV_nolen_const(_sv); })
#  else
#    define SvPVx_nolen_const(sv) (SvPV_nolen_const(sv))
#  endif
#endif

#ifndef PERL_ARGS_ASSERT_CK_WARNER
static void Perl_ck_warner(pTHX_ U32 err, const char* pat, ...);

#  ifdef vwarner
static
void
Perl_ck_warner(pTHX_ U32 err, const char* pat, ...)
{
  va_list args;

  PERL_UNUSED_ARG(err);
  if (ckWARN(err)) {
    va_list args;
    va_start(args, pat);
    vwarner(err, pat, &args);
    va_end(args);
  }
}
#  else
/* yes this replicates my_warner */
static
void
Perl_ck_warner(pTHX_ U32 err, const char* pat, ...)
{
  SV *sv;
  va_list args;

  PERL_UNUSED_ARG(err);

  va_start(args, pat);
  sv = vnewSVpvf(pat, &args);
  va_end(args);
  sv_2mortal(sv);
  warn("%s", SvPV_nolen(sv));
}
#  endif
#endif

#define PERL_VERSION_DECIMAL(r,v,s) (r*1000000 + v*1000 + s)
#define PERL_DECIMAL_VERSION \
	PERL_VERSION_DECIMAL(PERL_REVISION,PERL_VERSION,PERL_SUBVERSION)
#define PERL_VERSION_LT(r,v,s) \
	(PERL_DECIMAL_VERSION < PERL_VERSION_DECIMAL(r,v,s))
#define PERL_VERSION_GE(r,v,s) \
	(PERL_DECIMAL_VERSION >= PERL_VERSION_DECIMAL(r,v,s))

#if PERL_VERSION_LT(5,15,4)
#  define ISA_VERSION_OBJ(v) (sv_isobject(v) && sv_derived_from(v,"version"))
#else
#  define ISA_VERSION_OBJ(v) (sv_isobject(v) && sv_derived_from_pvn(v,"version",7,0))
#endif


#ifndef PERL_ARGS_ASSERT_CROAK_XS_USAGE
#define PERL_ARGS_ASSERT_CROAK_XS_USAGE assert(cv); assert(params)

/* prototype to pass -Wmissing-prototypes */
STATIC void
S_croak_xs_usage(pTHX_ const CV *const cv, const char *const params);

STATIC void
S_croak_xs_usage(pTHX_ const CV *const cv, const char *const params)
{
    const GV *const gv = CvGV(cv);

    PERL_ARGS_ASSERT_CROAK_XS_USAGE;

    if (gv) {
        const char *const gvname = GvNAME(gv);
        const HV *const stash = GvSTASH(gv);
        const char *const hvname = stash ? HvNAME(stash) : NULL;

        if (hvname)
            Perl_croak_nocontext("Usage: %s::%s(%s)", hvname, gvname, params);
        else
            Perl_croak_nocontext("Usage: %s(%s)", gvname, params);
    } else {
        /* Pants. I don't think that it should be possible to get here. */
        Perl_croak_nocontext("Usage: CODE(0x%"UVxf")(%s)", PTR2UV(cv), params);
    }
}

#ifdef PERL_IMPLICIT_CONTEXT
#define croak_xs_usage(a,b)	S_croak_xs_usage(aTHX_ a,b)
#else
#define croak_xs_usage		S_croak_xs_usage
#endif

#endif

#if PERL_VERSION_GE(5,9,0) && !defined(PERL_CORE)

#  define VUTIL_REPLACE_CORE 1

const char * Perl_scan_version2(pTHX_ const char *s, SV *rv, bool qv);
SV * Perl_new_version2(pTHX_ SV *ver);
SV * Perl_upg_version2(pTHX_ SV *sv, bool qv);
SV * Perl_vstringify2(pTHX_ SV *vs);
SV * Perl_vverify2(pTHX_ SV *vs);
SV * Perl_vnumify2(pTHX_ SV *vs);
SV * Perl_vnormal2(pTHX_ SV *vs);
SV * Perl_vstringify2(pTHX_ SV *vs);
int Perl_vcmp2(pTHX_ SV *lsv, SV *rsv);
const char * Perl_prescan_version2(pTHX_ const char *s, bool strict, const char** errstr, bool *sqv, int *ssaw_decimal, int *swidth, bool *salpha);

#  define SCAN_VERSION(a,b,c)	Perl_scan_version2(aTHX_ a,b,c)
#  define NEW_VERSION(a)	Perl_new_version2(aTHX_ a)
#  define UPG_VERSION(a,b)	Perl_upg_version2(aTHX_ a, b)
#  define VSTRINGIFY(a)		Perl_vstringify2(aTHX_ a)
#  define VVERIFY(a)		Perl_vverify2(aTHX_ a)
#  define VNUMIFY(a)		Perl_vnumify2(aTHX_ a)
#  define VNORMAL(a)		Perl_vnormal2(aTHX_ a)
#  define VCMP(a,b)		Perl_vcmp2(aTHX_ a,b)
#  define PRESCAN_VERSION(a,b,c,d,e,f,g)	Perl_prescan_version2(aTHX_ a,b,c,d,e,f,g)
#  undef is_LAX_VERSION
#  define is_LAX_VERSION(a,b) \
	(a != Perl_prescan_version2(aTHX_ a, FALSE, b, NULL, NULL, NULL, NULL))
#  undef is_STRICT_VERSION
#  define is_STRICT_VERSION(a,b) \
	(a != Perl_prescan_version2(aTHX_ a, TRUE, b, NULL, NULL, NULL, NULL))

#else

const char * Perl_scan_version(pTHX_ const char *s, SV *rv, bool qv);
SV * Perl_new_version(pTHX_ SV *ver);
SV * Perl_upg_version(pTHX_ SV *sv, bool qv);
SV * Perl_vverify(pTHX_ SV *vs);
SV * Perl_vnumify(pTHX_ SV *vs);
SV * Perl_vnormal(pTHX_ SV *vs);
SV * Perl_vstringify(pTHX_ SV *vs);
int Perl_vcmp(pTHX_ SV *lsv, SV *rsv);
const char * Perl_prescan_version(pTHX_ const char *s, bool strict, const char** errstr, bool *sqv, int *ssaw_decimal, int *swidth, bool *salpha);

#  define SCAN_VERSION(a,b,c)	Perl_scan_version(aTHX_ a,b,c)
#  define NEW_VERSION(a)	Perl_new_version(aTHX_ a)
#  define UPG_VERSION(a,b)	Perl_upg_version(aTHX_ a, b)
#  define VSTRINGIFY(a)		Perl_vstringify(aTHX_ a)
#  define VVERIFY(a)		Perl_vverify(aTHX_ a)
#  define VNUMIFY(a)		Perl_vnumify(aTHX_ a)
#  define VNORMAL(a)		Perl_vnormal(aTHX_ a)
#  define VCMP(a,b)		Perl_vcmp(aTHX_ a,b)

#  define PRESCAN_VERSION(a,b,c,d,e,f,g)	Perl_prescan_version(aTHX_ a,b,c,d,e,f,g)
#  ifndef is_LAX_VERSION
#    define is_LAX_VERSION(a,b) \
	(a != Perl_prescan_version(aTHX_ a, FALSE, b, NULL, NULL, NULL, NULL))
#  endif
#  ifndef is_STRICT_VERSION
#    define is_STRICT_VERSION(a,b) \
	(a != Perl_prescan_version(aTHX_ a, TRUE, b, NULL, NULL, NULL, NULL))
#  endif

#endif

#if PERL_VERSION_LT(5,11,4)
#  define BADVERSION(a,b,c) \
	if (b) { \
	    *b = c; \
	} \
	return a;

#  define PERL_ARGS_ASSERT_PRESCAN_VERSION	\
	assert(s); assert(sqv); assert(ssaw_decimal);\
	assert(swidth); assert(salpha);

#  define PERL_ARGS_ASSERT_SCAN_VERSION	\
	assert(s); assert(rv)
#  define PERL_ARGS_ASSERT_NEW_VERSION	\
	assert(ver)
#  define PERL_ARGS_ASSERT_UPG_VERSION	\
	assert(ver)
#  define PERL_ARGS_ASSERT_VVERIFY	\
	assert(vs)
#  define PERL_ARGS_ASSERT_VNUMIFY	\
	assert(vs)
#  define PERL_ARGS_ASSERT_VNORMAL	\
	assert(vs)
#  define PERL_ARGS_ASSERT_VSTRINGIFY	\
	assert(vs)
#  define PERL_ARGS_ASSERT_VCMP	\
	assert(lhv); assert(rhv)
#  define PERL_ARGS_ASSERT_CK_WARNER      \
	assert(pat)
#endif

/* from cperl */
#ifndef strNEc
#  define strNEc(s, c) memNE(s, ("" c ""), sizeof(c))
#endif

#if PERL_VERSION_LT(5,27,9)
#  define LC_NUMERIC_LOCK
#  define LC_NUMERIC_UNLOCK
#if PERL_VERSION_LT(5,19,0)
#   undef STORE_LC_NUMERIC_SET_STANDARD
#   undef RESTORE_LC_NUMERIC
#   undef DECLARATION_FOR_LC_NUMERIC_MANIPULATION
# ifdef USE_LOCALE
#    define DECLARATION_FOR_LC_NUMERIC_MANIPULATION char *loc
#    define STORE_NUMERIC_SET_STANDARD()\
	loc = savepv(setlocale(LC_NUMERIC, NULL));  \
	SAVEFREEPV(loc); \
	setlocale(LC_NUMERIC, "C");

#    define RESTORE_LC_NUMERIC()\
	setlocale(LC_NUMERIC, loc);
# else
#    define DECLARATION_FOR_LC_NUMERIC_MANIPULATION
#    define STORE_LC_NUMERIC_SET_STANDARD()
#    define RESTORE_LC_NUMERIC()
#   endif
# endif
#endif

#ifndef LOCK_NUMERIC_STANDARD
#  define LOCK_NUMERIC_STANDARD()
#endif
#ifndef UNLOCK_NUMERIC_STANDARD
#  define UNLOCK_NUMERIC_STANDARD()
#endif

/* The names of these changed in 5.28 */
#ifndef LOCK_LC_NUMERIC_STANDARD
#  define LOCK_LC_NUMERIC_STANDARD() LOCK_NUMERIC_STANDARD()
#endif
#ifndef UNLOCK_LC_NUMERIC_STANDARD
#  define UNLOCK_LC_NUMERIC_STANDARD() UNLOCK_NUMERIC_STANDARD()
#endif

/* ex: set ro: */


lib/version.pm
#!perl -w
package version;

use 5.006002;
use strict;
use warnings::register;

our $VERSION = '0.9923_02';
#$VERSION =~ s/c$//;
our $CLASS = 'version';
our (@ISA, $STRICT, $LAX);

# !!!!Delete this next block completely when adding to Perl core!!!!
{
    local $SIG{'__DIE__'};
    eval "use version::vxs $VERSION";
    if ( $@ ) { # don't have the XS version installed
	eval "use version::vpp $VERSION"; # don't tempt fate
	die "$@" if ( $@ );
	push @ISA, "version::vpp";
	local $^W;
	*version::qv = \&version::vpp::qv;
	*version::declare = \&version::vpp::declare;
	*version::_VERSION = \&version::vpp::_VERSION;
	*version::vcmp = \&version::vpp::vcmp;
	*version::new = \&version::vpp::new;
	*version::numify = \&version::vpp::numify;
	*version::normal = \&version::vpp::normal;
	if ($] >= 5.009000) {
	    no strict 'refs';
	    *version::stringify = \&version::vpp::stringify;
	    *{'version::(""'} = \&version::vpp::stringify;
	    *{'version::(<=>'} = \&version::vpp::vcmp;
	    *{'version::(cmp'} = \&version::vpp::scmp;
	    *version::parse = \&version::vpp::parse;
	}
    }
    else { # use XS module
	push @ISA, "version::vxs";
	local $^W;
	*version::declare = \&version::vxs::declare;
	*version::qv = \&version::vxs::qv;
	*version::_VERSION = \&version::vxs::_VERSION;
	*version::vcmp = \&version::vxs::VCMP;
	*version::new = \&version::vxs::new;
	*version::numify = \&version::vxs::numify;
	*version::normal = \&version::vxs::normal;
	if ($] >= 5.009000) {
	    no strict 'refs';
	    *version::stringify = \&version::vxs::stringify;
	    *{'version::(""'} = \&version::vxs::stringify;
	    *{'version::(<=>'} = \&version::vxs::VCMP;
	    *{'version::(cmp'} = \&version::vxs::VCMP;
	    *version::parse = \&version::vxs::parse;
	}
    }
}

# avoid using Exporter
require version::regex;
*version::is_lax = \&version::regex::is_lax;
*version::is_strict = \&version::regex::is_strict;
*LAX = \$version::regex::LAX;
*LAX_DECIMAL_VERSION = \$version::regex::LAX_DECIMAL_VERSION;
*LAX_DOTTED_DECIMAL_VERSION = \$version::regex::LAX_DOTTED_DECIMAL_VERSION;
*STRICT = \$version::regex::STRICT;
*STRICT_DECIMAL_VERSION = \$version::regex::STRICT_DECIMAL_VERSION;
*STRICT_DOTTED_DECIMAL_VERSION = \$version::regex::STRICT_DOTTED_DECIMAL_VERSION;

sub import {
    no strict 'refs';
    my ($class) = shift;

    # Set up any derived class
    unless ($class eq $CLASS) {
	local $^W;
	*{$class.'::declare'} =  \&{$CLASS.'::declare'};
	*{$class.'::qv'} = \&{$CLASS.'::qv'};
    }

    my %args;
    if (@_) { # any remaining terms are arguments
	map { $args{$_} = 1 } @_
    }
    else { # no parameters at all on use line
	%args =
	(
	    qv => 1,
	    'UNIVERSAL::VERSION' => 1,
	);
    }

    my $callpkg = caller();

    if (exists($args{declare})) {
	*{$callpkg.'::declare'} =
	    sub {return $class->declare(shift) }
	  unless defined(&{$callpkg.'::declare'});
    }

    if (exists($args{qv})) {
	*{$callpkg.'::qv'} =
	    sub {return $class->qv(shift) }
	  unless defined(&{$callpkg.'::qv'});
    }

    if (exists($args{'UNIVERSAL::VERSION'})) {
	local $^W;
	*UNIVERSAL::VERSION
		= \&{$CLASS.'::_VERSION'};
    }

    if (exists($args{'VERSION'})) {
	*{$callpkg.'::VERSION'} = \&{$CLASS.'::_VERSION'};
    }

    if (exists($args{'is_strict'})) {
	*{$callpkg.'::is_strict'} = \&{$CLASS.'::is_strict'}
	  unless defined(&{$callpkg.'::is_strict'});
    }

    if (exists($args{'is_lax'})) {
	*{$callpkg.'::is_lax'} = \&{$CLASS.'::is_lax'}
	  unless defined(&{$callpkg.'::is_lax'});
    }
}


1;


lib/version/regex.pm
package version::regex;

use strict;

use vars qw(
    $VERSION $CLASS $STRICT $LAX
    $STRICT_DECIMAL_VERSION $STRICT_DOTTED_DECIMAL_VERSION
    $LAX_DECIMAL_VERSION $LAX_DOTTED_DECIMAL_VERSION
);

$VERSION = '0.9921_01';

#--------------------------------------------------------------------------#
# Version regexp components
#--------------------------------------------------------------------------#

# Fraction part of a decimal version number.  This is a common part of
# both strict and lax decimal versions

my $FRACTION_PART = qr/\.[0-9]+/;

# First part of either decimal or dotted-decimal strict version number.
# Unsigned integer with no leading zeroes (except for zero itself) to
# avoid confusion with octal.

my $STRICT_INTEGER_PART = qr/0|[1-9][0-9]*/;

# First part of either decimal or dotted-decimal lax version number.
# Unsigned integer, but allowing leading zeros.  Always interpreted
# as decimal.  However, some forms of the resulting syntax give odd
# results if used as ordinary Perl expressions, due to how perl treats
# octals.  E.g.
#   version->new("010" ) == 10
#   version->new( 010  ) == 8
#   version->new( 010.2) == 82  # "8" . "2"

my $LAX_INTEGER_PART = qr/[0-9]+/;

# Second and subsequent part of a strict dotted-decimal version number.
# Leading zeroes are permitted, and the number is always decimal.
# Limited to three digits to avoid overflow when converting to decimal
# form and also avoid problematic style with excessive leading zeroes.

my $STRICT_DOTTED_DECIMAL_PART = qr/\.[0-9]{1,3}/;

# Second and subsequent part of a lax dotted-decimal version number.
# Leading zeroes are permitted, and the number is always decimal.  No
# limit on the numerical value or number of digits, so there is the
# possibility of overflow when converting to decimal form.

my $LAX_DOTTED_DECIMAL_PART = qr/\.[0-9]+/;

# Alpha suffix part of lax version number syntax.  Acts like a
# dotted-decimal part.

#my $LAX_ALPHA_PART = qr/_[0-9]+/;
my $LAX_ALPHA_PART = qr/(?:_[0-9]+c?|c?)/;

#--------------------------------------------------------------------------#
# Strict version regexp definitions
#--------------------------------------------------------------------------#

# Strict decimal version number.

$STRICT_DECIMAL_VERSION =
    qr/ $STRICT_INTEGER_PART $FRACTION_PART? c? /x;

# Strict dotted-decimal version number.  Must have both leading "v" and
# at least three parts, to avoid confusion with decimal syntax.

$STRICT_DOTTED_DECIMAL_VERSION =
    qr/ v $STRICT_INTEGER_PART $STRICT_DOTTED_DECIMAL_PART{2,} c? /x;

# Complete strict version number syntax -- should generally be used
# anchored: qr/ \A $STRICT \z /x

$STRICT =
    qr/ $STRICT_DECIMAL_VERSION | $STRICT_DOTTED_DECIMAL_VERSION /x;

#--------------------------------------------------------------------------#
# Lax version regexp definitions
#--------------------------------------------------------------------------#

# Lax decimal version number.  Just like the strict one except for
# allowing an alpha suffix or allowing a leading or trailing
# decimal-point

$LAX_DECIMAL_VERSION =
    qr/ $LAX_INTEGER_PART (?: $FRACTION_PART | \. )? $LAX_ALPHA_PART?
	|
	$FRACTION_PART $LAX_ALPHA_PART?
    /x;

# Lax dotted-decimal version number.  Distinguished by having either
# leading "v" or at least three non-alpha parts.  Alpha part is only
# permitted if there are at least two non-alpha parts. Strangely
# enough, without the leading "v", Perl takes .1.2 to mean v0.1.2,
# so when there is no "v", the leading part is optional

$LAX_DOTTED_DECIMAL_VERSION =
    qr/
	v $LAX_INTEGER_PART (?: $LAX_DOTTED_DECIMAL_PART+ $LAX_ALPHA_PART? )?
	|
	$LAX_INTEGER_PART? $LAX_DOTTED_DECIMAL_PART{2,} $LAX_ALPHA_PART?
    /x;

# Complete lax version number syntax -- should generally be used
# anchored: qr/ \A $LAX \z /x
#
# The string 'undef' is a special case to make for easier handling
# of return values from ExtUtils::MM->parse_version

$LAX =
    qr/ undef | $LAX_DOTTED_DECIMAL_VERSION | $LAX_DECIMAL_VERSION /x;

#--------------------------------------------------------------------------#

# Preloaded methods go here.
sub is_strict	{ defined $_[0] && $_[0] =~ qr/ \A $STRICT \z /x }
sub is_lax	{ defined $_[0] && $_[0] =~ qr/ \A $LAX \z /x }

1;


Makefile.PL
# This Makefile.PL stolen from Params::Validate
#
# The perl/C checking voodoo is stolen from Graham Barr's
# Scalar-List-Utils distribution.

use strict;

use Config qw(%Config);
use ExtUtils::MakeMaker;
use File::Temp qw/tempdir tempfile/;

if ($] < 5.006002) {
    die <<'EOL'
Perl 5.006002 required.  Please install version-0.9906 if you need
to support an earlier Perl release.
EOL
}

use vars qw ($no_xs $force_xs);
unlink 'pm_to_blib'; # belts and braces

if ($ENV{PERL_ONLY}) {
    $no_xs = 1;
}

for (@ARGV)
{
    /^--perl_only/ and $no_xs = 1;
    /^--perl-only/ and $no_xs = 1;
    /^--xs/ and $force_xs = 1;
}

if ($] < 5.010) { # support pure Perl only
    $no_xs = 1;
}

unless (defined $no_xs or $force_xs)
{
    check_for_compiler()
        or no_cc();

}

write_makefile();

sub write_makefile
{
    my %test_prereq = (
        'Test::More'            => 0.45,
        'File::Temp'            => 0.13,
        'parent'                => 0.221,
    );

    WriteMakefile( VERSION_FROM    => 'lib/version.pm',
                   NAME            => 'version',
                   LICENSE         => 'perl',
                   MIN_PERL_VERSION=> 5.006002,
                   TEST_REQUIRES   => \%test_prereq,
                   NORECURS        => $no_xs,
                   ABSTRACT    => 'Structured version objects',
                   AUTHOR      => 'John Peacock <jpeacock@cpan.org>',
                   # cperl support by Reini Urban
                   ( $] >= 5.009001 && $] < 5.011000 ?
                     ( INSTALLDIRS => 'perl' ) :
                     ()
                   ),
		   ( ($] < 5.012
		       && ! $ENV{PERL_NO_HIGHLANDER}
		       && ! ( $ENV{PERL_MM_OPT}
			   && $ENV{PERL_MM_OPT} =~ /(?:INSTALL_BASE|PREFIX)/ )
		       && ! grep { /INSTALL_BASE/ || /PREFIX/ } @ARGV ) ?
		     ( UNINST => 1 ) :
		     ()
		   ),
                   PM              =>
                       {'lib/version.pm'  => '$(INST_LIBDIR)/version.pm',
                        'lib/version.pod' => '$(INST_LIBDIR)/version.pod',
                        'lib/version/regex.pm'  =>
                            '$(INST_LIBDIR)/version/regex.pm',
                        'lib/version/Internals.pod' =>
                            '$(INST_LIBDIR)/version/Internals.pod',
                        'vperl/vpp.pm' => '$(INST_LIBDIR)/version/vpp.pm',
                        },
                   PL_FILES        => {},
                   C               => [],
		   ( $no_xs ?
		       () :
		       ( DIR       => ['vutil'])
		   ),
                   dist            => {
                       COMPRESS => 'gzip -9f',
                       SUFFIX => 'gz',
                   },
                   META_MERGE => {
                       "meta-spec" => { version => 2 },
                       resources => {
                           repository => {
                               type => 'git',
                               url => 'https://github.com/rurban/version'
                           },
                           bugtracker => {
                               web => 'https://rt.cpan.org/Public/Dist/Display.html?Name=version',
                               mailto => 'bug-version@rt.cpan.org',
                           },
                       },
                       no_index => {
                           package => ['charstar'],
                       },
                   },
                 );
}

sub no_cc
{
    $no_xs = 1;
    print <<'EOF';

 I cannot determine if you have a C compiler
 so I will install a perl-only implementation

 You can force installation of the XS version with

    perl Makefile.PL --xs

EOF

}

sub check_for_compiler
{
    # IMPORTANT NOTE:  This is NOT used to determine how to compile
    # extensions properly; EU::MM does that for us.  This is only
    # intended to see if that is likely to succeed.  We do not try
    # to do anything here except compile (not even link).  If you
    # want this to be a full featured test, feel free to submit a
    # patch or do something useful.

    print "Testing if you have a C compiler\n";

    eval { require ExtUtils::CBuilder };
    if ($@)
    {
        return _check_for_compiler_manually();
    }
    else
    {
        return _check_for_compiler_with_cbuilder();
    }
}

sub _check_for_compiler_with_cbuilder
{
    my $cb = ExtUtils::CBuilder->new( quiet => 1 );

    return $cb->have_compiler;
}

sub _check_for_compiler_manually
{
    unless ( open F, ">test.c" )
    {
        warn "Cannot write test.c, skipping test compilation and installing pure Perl version.\n";
        return 0;
    }

    print F <<'EOF';
int main() { return 0; }
EOF

    close F or return 0;

    my ($cc, $ccflags, $obj_ext) = map { $Config{$_} } qw/cc ccflags obj_ext/;

    my $command;
    if ($^O =~ /(dos|win32)/i && $Config{'cc'} =~ /^cl/) {
        $command = "$cc $ccflags /c test.c";
    }
    elsif ($Config{gccversion}) {
        $command = "$cc $ccflags -o test$obj_ext test.c";
    }
    else {
        warn "Unsupported system: can't test compiler availability. Patches welcome...";
        return 0;
    }

    my $retval = system( $command );
    map { unlink $_ if -f $_ } ('test.c',"test$obj_ext");

    return not($retval); # system returns -1
}
